paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/align/AlignedSourceReader.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.flink.util.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.flink.runtime.io.AvailabilityProvider.AVAILABLE;
+import static org.apache.paimon.flink.FlinkConnectorOptions.CheckpointAlignMode;
+
+/**
+ * A non-parallel {@link org.apache.flink.api.connector.source.SourceReader} that uses Flip-27 to
+ * achieve the same function as {@link org.apache.paimon.flink.source.operator.MonitorFunction}.
+ *
+ * <p>Send {@link Split} to the downstream task at paimon snapshot granularity, and provide two
+ * alignment modes: {@link CheckpointAlignMode#STRICTLY} and {@link CheckpointAlignMode#LOOSELY}.
+ *
+ * <ol>
+ *   <li>STRICTLY: only one paimon snapshot is processed within a checkpoint interval.
+ *   <li>LOOSELY: several paimon snapshots are processed within a checkpoint interval.
+ * </ol>
+ */
+@SuppressWarnings("unchecked")
+public abstract class AlignedSourceReader
+        implements ExternallyInducedSourceReader<Split, AlignedSourceSplit> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AlignedSourceReader.class);
+    private static final int MAX_PENDING_PLANS = 10;
+
+    private final ReadBuilder readBuilder;
+
+    private final FutureCompletingBlockingDeque<AlignedSourceSplit> pendingSplits;
+
+    private final TreeSet<Long> pendingCheckpoints;
+
+    private final TreeMap<Long, Long> nextSnapshotPerCheckpoint;
+
+    private final long scanInterval;
+
+    private final boolean emitSnapshotWatermark;
+
+    private final ScheduledExecutorService executors;
+
+    private Long snapshotIdByNextCheckpoint;
+
+    private boolean blocking;
+
+    private boolean endOfScan;
+
+    private CompletableFuture<Void> availabilityFuture;
+
+    private transient StreamTableScan scan;
+
+    public AlignedSourceReader(
+            ReadBuilder readBuilder,
+            long scanInterval,
+            boolean emitSnapshotWatermark,
+            ScheduledExecutorService executors) {
+        this.readBuilder = readBuilder;
+        this.pendingSplits = new FutureCompletingBlockingDeque<>();
+        this.pendingCheckpoints = new TreeSet<>();
+        this.availabilityFuture = (CompletableFuture<Void>) AVAILABLE;
+        this.nextSnapshotPerCheckpoint = new TreeMap<>();
+        this.scanInterval = scanInterval;
+        this.emitSnapshotWatermark = emitSnapshotWatermark;
+        this.executors = executors;
+        this.snapshotIdByNextCheckpoint = null;
+        this.blocking = false;
+        this.endOfScan = false;
+    }
+
+    @Override
+    public void start() {
+        this.scan = readBuilder.newStreamScan();
```

Comment:
Can splits just be send to here? Distributed scan I feel is still a little bit not so good from the design point of view

Reply:
If the splits are still scanned and assigned by the `Enumerator`, there is a problem: in the `strictly-align` mode, if the source table has not produced the corresponding snapshot at the time of the checkpoint, then the checkpoint will wait until the timeout.

Reply:
But we can also send a event to reader to show that there is no split.

Reply:
So what should the reader do for this checkpoint? directly abort? This may cause unnecessary checkpoint failures.